stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -211,12 +211,20 @@ end
 
 function rmul!(A::AbstractMatrix, D::Diagonal)
     require_one_based_indexing(A)
+    nA, nD = size(A, 2), length(D.diag)
+    if nA != nD
+        throw(DimensionMismatch("second dimension of A, $nA, does not match the first of D, $nD"))
+    end
```

Comment:
I wonder if we should have a `matmul_boundscheck` function that we can use for all the matmul methods.

Reply:
That might be a good idea, but I'd rather call it something like `matmul_compat`, because it doesn't quite feel like bounds checking, but rather like size compatibility checking. Do you mind opening a separate issue for that? Someone would need to go through the entire stdlib with that function.

Reply:
I wondered this too. It could perhaps also absorb most of the `require_one_based_indexing(A)` calls.

Reply:
We have something like that in `LinearMaps.jl`, with the error message basically copied from `LinearAlgebra`.